virt-manager-0.9.4/man/virt-manager.1
.\" Automatically generated by Pod::Man 2.25 (Pod::Simple 3.16)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "VIRT-MANAGER 1"
.TH VIRT-MANAGER 1 "2012-07-30" "" "Virtual Machine Manager"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
virt\-manager \- display the virtual machine desktop management tool
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBvirt-manager\fR [\s-1OPTIONS\s0]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBvirt-manager\fR is a desktop tool for managing virtual machines. It
provides the ability to control the lifecycle of existing machines
(bootup/shutdown,pause/resume,suspend/restore), provision new virtual
machines and various types of store, manage virtual networks,
access the graphical console of virtual machines, and view performance
statistics, all done locally or remotely.
.SH "OPTIONS"
.IX Header "OPTIONS"
The following options are accepted when running \f(CW\*(C`virt\-manager\*(C'\fR:
.IP "\-h, \-help" 4
.IX Item "-h, -help"
Display command line help summary
.IP "\-c \s-1URI\s0, \-\-connect=URI" 4
.IX Item "-c URI, --connect=URI"
Specify the hypervisor connection \s-1URI\s0
.IP "\-\-debug" 4
.IX Item "--debug"
List debugging output to the console (normally this is only logged in
~/.virt\-manager/virt\-manager.log). This function implies \-\-no\-fork.
.IP "\-\-no\-fork" 4
.IX Item "--no-fork"
Don't fork \f(CW\*(C`virt\-manager\*(C'\fR off into the backround: run it blocking the
current terminal. Useful for seeing possible errors dumped to stdout/stderr.
.IP "\-\-no\-dbus" 4
.IX Item "--no-dbus"
Disable the DBus \s-1API\s0 used for remote control of the \f(CW\*(C`virt\-manager\*(C'\fR \s-1UI\s0
.IP "\-\-no\-conn\-autostart" 4
.IX Item "--no-conn-autostart"
Don't autostart any libvirt connections when launching \f(CW\*(C`virt\-manager\*(C'\fR.
.IP "\-\-show\-domain\-creator" 4
.IX Item "--show-domain-creator"
Display the wizard for creating new virtual machines
.IP "\-\-show\-domain\-editor=UUID" 4
.IX Item "--show-domain-editor=UUID"
Display the dialog for editing properties of the virtual machine with
unique \s-1ID\s0 matching \f(CW\*(C`UUID\*(C'\fR
.IP "\-\-show\-domain\-performance=UUID" 4
.IX Item "--show-domain-performance=UUID"
Display the dialog for monitoring performance of the virtual machine with
unique \s-1ID\s0 matching \f(CW\*(C`UUID\*(C'\fR
.IP "\-\-show\-domain\-console=UUID" 4
.IX Item "--show-domain-console=UUID"
Display the virtual console of the virtual machine with
unique \s-1ID\s0 matching \f(CW\*(C`UUID\*(C'\fR
.IP "\-\-show\-host\-summary" 4
.IX Item "--show-host-summary"
Display the main window summarizing performance for all virtual machines
on the host.
.SH "AUTHOR"
.IX Header "AUTHOR"
Written by Daniel P. Berrange and Cole Robinson.
.SH "BUGS"
.IX Header "BUGS"
Please see http://virt\-manager.org/page/BugReporting <http://virt-manager.org/page/BugReporting>
.SH "COPYRIGHT"
.IX Header "COPYRIGHT"
Copyright (C) 2006\-2007 Red Hat, Inc, and various contributors.
This is free software. You may redistribute copies of it under the terms of the \s-1GNU\s0 General
Public License \f(CW\*(C`http://www.gnu.org/licenses/gpl.html\*(C'\fR. There is \s-1NO\s0 \s-1WARRANTY\s0, to the extent
permitted by law.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\f(CWvirsh(1)\fR, \f(CW\*(C`virt\-viewer(1)\*(C'\fR, the project website \f(CW\*(C`http://virt\-manager.org\*(C'\fR


virt-manager-0.9.4/src/virt-manager.py
#
# Copyright (C) 2006 Red Hat, Inc.
# Copyright (C) 2006 Daniel P. Berrange <berrange@redhat.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
# MA 02110-1301 USA.
#

import os
import os.path
import sys

import logging
import traceback
import signal
import optparse

try:
    # Make sure we have a default '_' implementation, in case something
    # fails before gettext is set up
    __builtins__._ = lambda msg: msg
except:
    pass

def split_list(commastr):
    return [d for d in commastr.split(",") if d]

# These are substituted into code based on --prefix given to configure
appname = "virt-manager"
appversion = "0.9.4"
gettext_app = "virt-manager"
gettext_dir = "/usr/local/share/locale"
virtinst_str = "0.600.3"

asset_dir = "/usr/local/share/virt-manager"
ui_dir = asset_dir
icon_dir = asset_dir + "/icons"
pylib_dir = "/usr/local/share/virt-manager"
pyarchlib_dir = "/usr/local/lib/virt-manager"

default_qemu_user = "root"
rhel_enable_unsupported_opts = bool(int("1"))
preferred_distros = split_list("")

hv_packages = split_list("::KVM_PACKAGES::")
askpass_package = split_list("")
libvirt_packages = split_list("")

logging_setup = False

def _show_startup_error(msg, details):
    from virtManager.error import vmmErrorDialog
    err = vmmErrorDialog()
    title = _("Error starting Virtual Machine Manager")
    err.show_err(title + ": " + msg,
                 details=details,
                 title=title,
                 async=False,
                 debug=False)

def setup_pypath():
    global ui_dir, icon_dir

    # Hack to find assets in local dir for dev purposes
    if os.path.exists(os.getcwd() + "/src/virt-manager.py.in"):
        ui_dir = os.path.join(os.getcwd(), "src")
        icon_dir = os.path.join(os.getcwd(), "icons")
    else:
        sys.path.insert(0, pylib_dir)
        sys.path.insert(0, pyarchlib_dir)


def drop_tty():
    # We fork and setsid so that we drop the controlling
    # tty. This prevents libvirt's SSH tunnels from prompting
    # for user input if SSH keys/agent aren't configured.
    if os.fork() != 0:
        os._exit(0)

    os.setsid()

def drop_stdio():
    # We close STDIN/OUT/ERR since they're generally spewing
    # junk to console when domains are in process of shutting
    # down. Real errors will (hopefully) all be logged to the
    # main log file. This is also again to stop SSH prompting
    # for input
    for fd in range(0, 2):
        try:
            os.close(fd)
        except OSError:
            pass

    os.open(os.devnull, os.O_RDWR)
    os.dup2(0, 1)
    os.dup2(0, 2)

def parse_commandline():
    optParser = optparse.OptionParser(version=appversion,
                                      usage="virt-manager [options]")

    optParser.set_defaults(uuid=None)

    # Generate runtime performance profile stats with hotshot
    optParser.add_option("--profile", dest="profile",
        help=optparse.SUPPRESS_HELP, metavar="FILE")

    optParser.add_option("-c", "--connect", dest="uri",
        help="Connect to hypervisor at URI", metavar="URI")
    optParser.add_option("--debug", action="store_true", dest="debug",
        help="Print debug output to stdout (implies --no-fork)",
        default=False)
    optParser.add_option("--no-dbus", action="store_true", dest="nodbus",
        help="Disable DBus service for controlling UI")
    optParser.add_option("--no-fork", action="store_true", dest="nofork",
        help="Don't fork into background on startup")
    optParser.add_option("--no-conn-autostart", action="store_true",
                         dest="no_conn_auto",
                         help="Do not autostart connections")

    optParser.add_option("--show-domain-creator", action="callback",
        callback=opt_show_cb, dest="show",
        help="Show 'New VM' wizard")
    optParser.add_option("--show-domain-editor", type="string",
        metavar="UUID", action="callback", callback=opt_show_cb,
        help="Show domain details window")
    optParser.add_option("--show-domain-performance", type="string",
        metavar="UUID", action="callback", callback=opt_show_cb,
        help="Show domain performance window")
    optParser.add_option("--show-domain-console", type="string",
        metavar="UUID", action="callback", callback=opt_show_cb,
        help="Show domain graphical console window")
    optParser.add_option("--show-host-summary", action="callback",
       callback=opt_show_cb, help="Show connection details window")

    return optParser.parse_args()


def launch_specific_window(engine, show, uri, uuid):
    if not show:
        return

    logging.debug("Launching requested window '%s'", show)
    if show == 'creator':
        engine.show_domain_creator(uri)
    elif show == 'editor':
        engine.show_domain_editor(uri, uuid)
    elif show == 'performance':
        engine.show_domain_performance(uri, uuid)
    elif show == 'console':
        engine.show_domain_console(uri, uuid)
    elif show == 'summary':
        engine.show_host_summary(uri)

def _conn_state_changed(conn, engine, show, uri, uuid):
    if conn.state == conn.STATE_DISCONNECTED:
        return True
    if conn.state != conn.STATE_ACTIVE:
        return

    launch_specific_window(engine, show, uri, uuid)
    return True

# maps --show-* to engine (ie local instance) methods
def show_engine(engine, show, uri, uuid, no_conn_auto):
    conn = None

    # Do this regardless
    engine.show_manager()

    if uri:
        conn = engine.add_conn(uri)

        if conn and show:
            conn.connect_opt_out("state-changed",
                                 _conn_state_changed,
                                 engine, show, uri, uuid)

        engine.connect_to_uri(uri)

    if not no_conn_auto:
        engine.autostart_conns()

# maps --show-* to remote manager (ie dbus call) methods
def show_remote(managerObj, show, uri, uuid):
    # Do this regardless
    managerObj.show_manager()

    if show or uri or uuid:
        launch_specific_window(managerObj, show, uri, uuid)

def dbus_config(engine):
    """
    Setup dbus interface
    """
    import dbus
    from virtManager.remote import vmmRemote
    bus = None

    if os.getenv("DBUS_STARTER_ADDRESS") is None:
        bus = dbus.SessionBus()
    else:
        bus = dbus.StarterBus()

    dbusProxy = bus.get_object("org.freedesktop.DBus", "/org/freedesktop/DBus")
    dbusObj = dbus.Interface(dbusProxy, "org.freedesktop.DBus")

    if dbusObj.NameHasOwner("com.redhat.virt.manager"):
        # We're already running, so just talk to existing process
        managerProxy = bus.get_object("com.redhat.virt.manager",
                                      "/com/redhat/virt/manager")
        managerObj = dbus.Interface(managerProxy, "com.redhat.virt.manager")
        return managerObj

    else:
        # Grab the service to allow others to talk to us later
        name = dbus.service.BusName("com.redhat.virt.manager", bus=bus)
        vmmRemote(engine, name)

# Generic OptionParser callback for all --show-* options
# This routine stores UUID to options.uuid for all --show-* options
# where is metavar="UUID" and also sets options.show
def opt_show_cb(option, opt_str, value, parser):
    if option.metavar == "UUID":
        setattr(parser.values, "uuid", value)
    s = str(option)
    show = s[s.rindex('-') + 1:]
    setattr(parser.values, "show", show)

# Run me!
def main():
    setup_pypath()

    import virtManager
    from virtManager import cli

    cli.setup_i18n(gettext_app, gettext_dir)

    # Need to do this before GTK strips args like --sync
    gtk_error = None
    origargs = " ".join(sys.argv[:])

    # Urgh, pygtk merely logs a warning when failing to open
    # the X11 display connection, and lets everything carry
    # on as if all were fine. Ultimately bad stuff happens,
    # so lets catch it here & get the hell out...
    import warnings
    warnings.filterwarnings('error', module='gtk', append=True)
    try:
        import gobject

        # Set program name for gnome shell (before importing gtk, which
        # seems to call set_prgname on its own)
        if hasattr(gobject, "set_prgname"):
            gobject.set_prgname(appname)

        import gtk
        globals()["gtk"] = gtk
    except Warning, e:
        # ...the risk is we catch too much though
        # Damned if we do, damned if we dont :-)(
        gtk_error = str(e)
    except Exception, e:
        gtk_error = e
    warnings.resetwarnings()

    # Need to parse CLI after import gtk, since gtk strips --sync
    (options, ignore) = parse_commandline()

    # Only raise this error after parsing the CLI, so users at least
    # get --help output and CLI validation
    if gtk_error:
        if type(gtk_error) is str:
            raise RuntimeError(_("Unable to initialize GTK: %s") % gtk_error)
        raise gtk_error

    if not hasattr(gtk, "Builder"):
        raise RuntimeError("virt-manager requires GtkBuilder support. "
                           "Your gtk version appears to be too old.")

    cli.setup_logging(appname, options.debug)
    global logging_setup
    logging_setup = True

    logging.debug("Launched as: %s", origargs)
    logging.debug("GTK version: %s", str(gtk.gtk_version))
    logging.debug("virt-manager version: %s", appversion)
    logging.debug("virtManager import: %s", str(virtManager))

    cli.check_virtinst_version(virtinst_str)

    # Add our icon dir to icon theme
    icon_theme = gtk.icon_theme_get_default()
    icon_theme.prepend_search_path(icon_dir)

    gobject.threads_init()

    import dbus
    import dbus.mainloop.glib
    dbus.mainloop.glib.DBusGMainLoop(set_as_default=True)
    dbus.mainloop.glib.threads_init()
    import dbus.service

    # Specifically init config/gconf before the fork, so that pam
    # doesn't think we closed the app, therefor robbing us of
    # display access
    import virtManager.config
    import virtManager.util
    config = virtManager.config.vmmConfig(appname, appversion, ui_dir)
    virtManager.util.running_config = config
    config.default_qemu_user = default_qemu_user
    config.rhel6_defaults = rhel_enable_unsupported_opts
    config.preferred_distros = preferred_distros

    config.hv_packages = hv_packages
    config.libvirt_packages = libvirt_packages
    config.askpass_package = askpass_package

    import virtManager.guidiff
    virtManager.guidiff.is_gui(True)

    # Now we've got basic environment up & running we can fork
    if not options.nofork and not options.debug:
        drop_tty()
        drop_stdio()

        # Ignore SIGHUP, otherwise a serial console closing drops the whole app
        signal.signal(signal.SIGHUP, signal.SIG_IGN)

    from virtManager.engine import vmmEngine

    gtk.window_set_default_icon_name(appname)

    if options.show and options.uri == None:
        raise optparse.OptionValueError("can't use --show-* options "
                                        "without --connect")

    engine = vmmEngine()

    if not options.nodbus:
        try:
            managerObj = dbus_config(engine)
            if managerObj:
                # yes, we exit completely now - remote service is in charge
                logging.debug("Connected to already running instance.")
                show_remote(managerObj, options.show,
                            options.uri, options.uuid)
                return
        except:
            # Something went wrong doing dbus setup, just ignore & carry on
            logging.exception("Could not get connection to session bus, "
                              "disabling DBus service")

    # Hook libvirt events into glib main loop
    import virtManager.libvirtglib
    virtManager.libvirtglib.register_event_impl()

    # At this point we're either starting a brand new controlling instance,
    # or the dbus comms to existing instance has failed

    # Finally start the app for real
    show_engine(engine, options.show, options.uri, options.uuid,
                options.no_conn_auto)
    if options.profile != None:
        import hotshot
        prof = hotshot.Profile(options.profile)
        prof.runcall(gtk.main)
        prof.close()
    else:
        gtk.main()

if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        logging.debug("Received KeyboardInterrupt. Exiting application.")
    except SystemExit:
        raise
    except Exception, run_e:
        if logging_setup:
            logging.exception(run_e)
        if "gtk" not in globals():
            raise
        _show_startup_error(str(run_e), "".join(traceback.format_exc()))
